azavea/raster-foundry
# Add Tool Category route

## Overview
Add a tool category route for searching, listing, creating, and deleting tool categories

### Checklist


### Demo
Fetching categories
```
curl 'http://localhost:9091/api/tool-categories/' -H 'Authorization: Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJodHRwczovL3Jhc3Rlci1mb3VuZHJ5LmF1dGgwLmNvbS8iLCJzdWIiOiJnb29nbGUtb2F1dGgyfDExNjc1Mzk0Mzc5OTE5ODAyNzg2OSIsImF1ZCI6IlphQ3h5aWtpRExGTEtCT2ZQQjVSMzB0bmM4cjA2bnhVIiwiZXhwIjoxNDgxMDY1NTg1LCJpYXQiOjE0ODEwMjk1ODV9.Hd8MqEyPCGUvg38QQA6E2jV0jF3zUJbkvi3COwvqg0o'

{
  "count": 2,
  "pageSize": 30,
  "hasPrevious": false,
  "page": 0,
  "results": [{
    "modifiedAt": "2016-12-06T13:57:57.417Z",
    "createdAt": "2016-12-06T13:57:57.417Z",
    "createdBy": "google-oauth2|116753943799198027869",
    "category": "Raster Analysis",
    "slugLabel": "raster-analysis",
    "modifiedBy": "google-oauth2|116753943799198027869"
  }, {
    "modifiedAt": "2016-12-06T13:59:09.503Z",
    "createdAt": "2016-12-06T13:59:09.503Z",
    "createdBy": "google-oauth2|116753943799198027869",
    "category": "Test Imagery",
    "slugLabel": "test-imagery",
    "modifiedBy": "google-oauth2|116753943799198027869"
  }],
  "hasNext": false
}%
```
Creating an invalid category
```
> $ echo '{"category": "好"}' | curl -H "Content-Type: application/json" -X POST -d @- http://localhost:9091/api/tool-categories -H 'Authorization: Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJodHRwczovL3Jhc3Rlci1mb3VuZHJ5LmF1dGgwLmNvbS8iLCJzdWIiOiJnb29nbGUtb2F1dGgyfDExNjc1Mzk0Mzc5OTE5ODAyNzg2OSIsImF1ZCI6IlphQ3h5aWtpRExGTEtCT2ZQQjVSMzB0bmM4cjA2bnhVIiwiZXhwIjoxNDgxMTU2OTkzLCJpYXQiOjE0ODExMjA5OTN9.U7le-VlE78lgrcTdZYhD4Xn_hB_g3O6qyVighkA8VhE'

Invalid category: 好. Cannot slugify% 
```

Create category with unicode characters: categories can be created with latin / non-letter symbols as long as converting diacritics can be converted to ascii letters and removing symbols doesn't result in a non-unique slug label
```
> $ echo '{"category": "niche à chien"}' | curl -H "Content-Type: application/json" -X POST -d @- http://localhost:9091/api/tool-categories -H 'Authorization: Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJodHRwczovL3Jhc3Rlci1mb3VuZHJ5LmF1dGgwLmNvbS8iLCJzdWIiOiJnb29nbGUtb2F1dGgyfDExNjc1Mzk0Mzc5OTE5ODAyNzg2OSIsImF1ZCI6IlphQ3h5aWtpRExGTEtCT2ZQQjVSMzB0bmM4cjA2bnhVIiwiZXhwIjoxNDgxMTU2OTkzLCJpYXQiOjE0ODExMjA5OTN9.U7le-VlE78lgrcTdZYhD4Xn_hB_g3O6qyVighkA8VhE'
{
  "modifiedAt": "2016-12-03T08:00:12.004Z",
  "createdAt": "2016-12-03T08:00:12.004Z",
  "createdBy": "google-oauth2|116753943799198027869",
  "category": "niche à chien",
  "slugLabel": "niche-a-chien",
  "modifiedBy": "google-oauth2|116753943799198027869"
}%  
```

Searching for a category:
```
lucien@towamencin ~                                                                                                                                                                                    [9:47:37] 
> $ curl 'http://localhost:9091/api/tool-categories/' -H 'Authorization: Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJodHRwczovL3Jhc3Rlci1mb3VuZHJ5LmF1dGgwLmNvbS8iLCJzdWIiOiJnb29nbGUtb2F1dGgyfDExNjc1Mzk0Mzc5OTE5ODAyNzg2OSIsImF1ZCI6IlphQ3h5aWtpRExGTEtCT2ZQQjVSMzB0bmM4cjA2bnhVIiwiZXhwIjoxNDgxMTU2OTkzLCJpYXQiOjE0ODExMjA5OTN9.U7le-VlE78lgrcTdZYhD4Xn_hB_g3O6qyVighkA8VhE' 
{
  "count": 2,
  "pageSize": 30,
  "hasPrevious": false,
  "page": 0,
  "results": [{
    "modifiedAt": "2016-12-03T08:00:12.004Z",
    "createdAt": "2016-12-03T08:00:12.004Z",
    "createdBy": "google-oauth2|116753943799198027869",
    "category": "niche à chien",
    "slugLabel": "niche-a-chien",
    "modifiedBy": "google-oauth2|116753943799198027869"
  }, {
    "modifiedAt": "2016-12-03T08:05:01.887Z",
    "createdAt": "2016-12-03T08:05:01.887Z",
    "createdBy": "google-oauth2|116753943799198027869",
    "category": "Test",
    "slugLabel": "test",
    "modifiedBy": "google-oauth2|116753943799198027869"
  }],
  "hasNext": false
}%                                                                                                                                                                                                                
lucien@towamencin ~                                                                                                                                                                                    [9:47:46] 
> $ curl 'http://localhost:9091/api/tool-categories/?search=nich' -H 'Authorization: Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJodHRwczovL3Jhc3Rlci1mb3VuZHJ5LmF1dGgwLmNvbS8iLCJzdWIiOiJnb29nbGUtb2F1dGgyfDExNjc1Mzk0Mzc5OTE5ODAyNzg2OSIsImF1ZCI6IlphQ3h5aWtpRExGTEtCT2ZQQjVSMzB0bmM4cjA2bnhVIiwiZXhwIjoxNDgxMTU2OTkzLCJpYXQiOjE0ODExMjA5OTN9.U7le-VlE78lgrcTdZYhD4Xn_hB_g3O6qyVighkA8VhE'
{
  "count": 1,
  "pageSize": 30,
  "hasPrevious": false,
  "page": 0,
  "results": [{
    "modifiedAt": "2016-12-03T08:00:12.004Z",
    "createdAt": "2016-12-03T08:00:12.004Z",
    "createdBy": "google-oauth2|116753943799198027869",
    "category": "niche à chien",
    "slugLabel": "niche-a-chien",
    "modifiedBy": "google-oauth2|116753943799198027869"
  }],
  "hasNext": false
}%    
```

### Notes

For now, the category endpoint isn't user role limited, but category creation / deletion should be role restricted to admins / removed entirely and managed through migrations in the future.

## Testing Instructions

 * Use curl  or another request tool to do the following:
  * Create a category with only ASCII characters
  * Create a category with whitespace, dashes, and misc symbols - it should slugify, but the category will be searchable using the symbols)
  * Try to create a category which consists only of non-latin characters (eg: translate dog to chinese and use the result to create a category)
  * Create a category which contains diacritical marks (see demo examples)
  * Search for categories by the display string (Category attribute) where it contains diacritical marks - it should work
